### DISCUSSION TITLE
Creating a Python virtualenv from the commandline

### DISCUSSION THREAD

**Discussion by janbuchar** [jdx/mise#2492] [Q&A]
I know I can kickstart a "project" with `mise use python`. Can I also set up a virtualenv with `mise use`? So that the resulting `.mise.toml` looks like this?

```toml
[tools]
python = {version = "latest", virtualenv = ".venv"}
```

**Comment by mcallaway** [comment]
I'm interested in this topic as well. I want to set up a new python project with "normal" use of commands, ie. not "manually" editing a `.mise.toml` file with an editor. I also want to use the `poetry` plugin. To date I've been doing what I want by using `cat` with a HEREDOC. Like this:

```
❯ git init newproj
Initialized empty Git repository in /Users/me/git/newproj/.git/

❯ cd newproj

❯ type mise # Here I'm just indicating that I've activated mise via .zshrc
mise is a shell function

❯ cat > .mise.toml <<EOF
[tools]
poetry = {version='latest', pyproject='pyproject.toml'}
python = {version='3.12', virtualenv='.venv'}
EOF
mise setting up virtualenv at: /Users/me/git/newproj/.venv
mise-poetry: No pyproject.toml found. Execute `poetry init` to create `/Users/me/git/newproj/pyproject.toml` first.
```

I think it would be better to be able to say something like:

    mise use python@3.12,virtualenv=.venv

It's not clear to me how plugins should be included in command line parsing however.

**Reply by jdx** [reply]
I mentioned this yesterday in another thread but I would just use a toml editor if it were me: https://github.com/gnprice/toml-cli

```
mise use python@3.12,virtualenv=.venv
```

this syntax should not be the way forward since having the venv as part of the python tool is legacy behavior I may remove one day in favor of the env directive.

Also note there is the [`python-venv-auto-create`](https://mise.jdx.dev/lang/python.html#python-venv-auto-create) setting.

**Reply by janbuchar** [reply]
I mean, yeah, just calling `toml set` would probably do it, but still, something like `mise set virtualenv` with a reasonable default would be very useful.

**Reply by jdx** [reply]
or hell, maybe `mise toml add env._python.virtualenv .venv`

**Reply by jdx** [reply]
I know that's more verbose, but it would allow editing of anything in the config files

**Reply by janbuchar** [reply]
Wow, I like that idea! It's admittedly very close to just using toml-cli, but you don't have to install another tool

**Reply by jdx** [reply]
Go ahead and file a feature request, I'd like to see it too

EDIT: https://github.com/jdx/mise/issues/2521

**Reply by Marukome0743** [reply]
Following up on this older discussion: the config-editing command proposed in this thread was implemented in [#2657](https://github.com/jdx/mise/pull/2657) and is now available as `mise config set`.

The Python tool’s legacy `virtualenv` option is no longer the recommended configuration. The current CLI-only workflow is:

```sh
mise use python@3.12
mise config set env._.python.venv.path .venv
mise config set env._.python.venv.create true -t bool
```

This produces the current recommended form without manually editing `mise.toml`:

```toml
[tools]
python = "3.12"

[env._.python.venv]
path = ".venv"
create = true
```

I verified these commands against the current main branch and confirmed that the resulting config is parsed and written correctly.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
